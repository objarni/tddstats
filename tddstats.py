import colors
import time

RED, GREEN, BLUE = range(3)


class TDDSession:
    def __init__(self, timer=None):
        self.timer = timer or time.time
        self.session_log = None

    def total_time(self):
        return sum(t[1] for t in self.session_log.get_log())

    def click(self):
        if self.session_log is None:
            self.session_log = SessionLog(start_time=self.timer())
        else:
            self.session_log.switch(time=self.timer())

    def sum_red_green_blue(self):
        log = self.session_log.get_log()
        sums = [sum(t[1] for t in log
                    if t[0] == value)
                for value in range(3)]
        return tuple(sums)

    def percents(self):
        total = self.total_time()
        percents = [int(value / total * 100)
                    for value in self.sum_red_green_blue()]
        return tuple(percents)

    def report(self):
        (r, g, b) = self.sum_red_green_blue()
        (pr, pg, pb) = self.percents()
        return f"""--- SESSION SUMMARY ---
Total time:        {self.total_time():1.1f} s
Adding tests:      {b:1.1f} s ({pb}%)
Making them pass:  {r:1.1f} s ({pr}%)
Refactoring:       {g:1.1f} s ({pg}%)
"""


class SessionLog:
    def __init__(self, start_time):
        self.mode = BLUE
        self.time = start_time
        self.log = []

    def switch(self, time):
        dt = time - self.time
        self.time = time
        self.log.append((self.mode, dt))
        self.mode = (self.mode + 1) % 3

    def get_log(self):
        return self.log


if __name__ == '__main__':
    session = TDDSession()
    print("TDD stats session.")
    print("Press enter to start session (in blue, add a test).")
    print("Press enter to switch state (to red, then green, then blue again etc.).")
    print("Type Q and press enter at any time to stop session.")
    input("[waiting for first enter]")
    mode = BLUE
    mode_names = "BLUE RED GREEN".lower().split()
    mode_desc = ["add a test, the fox says", "quickly make all tests pass, says the rabbit and runs", "enjoy refactoring, cat meows"]
    while True:
        session.click()
        mode = (mode + 1) % 3
        color = mode_names[mode]
        desc = mode_desc[mode]
        print(colors.color(f"In mode: {color} ({desc})", bg=color))
        ans = input().upper()
        if ans == "Q":
            print(session.report())
            break
