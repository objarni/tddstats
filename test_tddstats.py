from pytest import approx
from approvaltests import verify

''' SCENARIOS
green paths
2 cykler, lika stora slices:
  användaren klickar förbi 2 hela cykler och sedan exitar,
  med 1.1 sekund i varje mode -->
  Session lasted 6.6 seconds:
  - RED:   33% of time (2.2 seconds)
  - GREEN: 33% of time (2.2 seconds)
  - BLUE:  33% of time (2.2 seconds)

1 cykel, olika stora slices:
  användaren lägger 3 sekunder i varje red, 2 i green, 1 i blue
  Session lasted 6 seconds:
  - RED:   50% of time (3 seconds)
  - GREEN: 34% of time (2 seconds)
  - BLUE:  17% of time (1 seconds)

'''
from tddstats import TDDSession, SessionLog, RED, GREEN, BLUE


def test_2_cycle_same_lengths():
    times = [1.1 * i for i in range(8)]

    def fake_timer():
        t = fake_timer.times[0]
        fake_timer.times = fake_timer.times[1:]
        return t

    fake_timer.times = times

    session = TDDSession(fake_timer)
    for _ in range(7):
        session.click()

    verify(session.report())


def test_1_cycle_different_lengths():
    times = [0, 3, 5, 6]

    def fake_timer():
        t = fake_timer.times[0]
        fake_timer.times = fake_timer.times[1:]
        return t

    fake_timer.times = times

    session = TDDSession(timer=fake_timer)
    session.click()
    session.click()
    session.click()
    session.click()

    verify(session.report())


def test_session_one_switch():
    session_log = SessionLog(start_time=2)
    session_log.switch(time=3)
    assert [(BLUE, 1)] == session_log.get_log()


def test_session_two_switches():
    session_log = SessionLog(start_time=2)
    session_log.switch(time=3)
    session_log.switch(time=5.5)
    assert [(BLUE, 1), (RED, 2.5)] == session_log.get_log()


def test_session_three_switches():
    session_log = SessionLog(start_time=0)
    session_log.switch(time=1)
    session_log.switch(time=2)
    session_log.switch(time=3)
    assert [(BLUE, 1), (RED, 1), (GREEN, 1)] == session_log.get_log()
