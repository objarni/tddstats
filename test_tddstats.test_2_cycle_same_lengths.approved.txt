--- SESSION SUMMARY ---
Total time:        6.6 s
Adding tests:      2.2 s (33%)
Making them pass:  2.2 s (33%)
Refactoring:       2.2 s (33%)
